File: 122.买卖股票的最佳时机-ii.js
/*
 * @lc app=leetcode.cn id=122 lang=javascript
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let dp = [[0, -prices[0]]];
  for (let i = 1; i < prices.length; i++) {
    dp[i] = [0, 0];
    dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
    dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
  }
  return dp[prices.length - 1][0];
};

const maxProfit1 = (prices) => {
  const dp = [[0, -prices[0]]];
  for (let i = 1; i < prices.length; i++) {
    dp[i] = [0, 0];
    dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
    dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
  }
  return dp[prices.length - 1][0];
};

console.log(maxProfit([7, 1, 5, 3, 6, 4]));
console.log(maxProfit1([7, 1, 5, 3, 6, 4]));

// @lc code=end

File: 122.买卖股票的最佳时机-ii.go
package leetcode122

/*
 * @lc app=leetcode.cn id=122 lang=golang
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start
func maxProfit2(prices []int) int {
	if len(prices) < 2 {
		return 0
	}
	res := 0

	for i := 1; i < len(prices); i++ {
		diff := prices[i] - prices[i-1]
		if diff > 0 {
			res += diff
		}
	}
	return res
}

func maxProfit(prices []int) int {
	dp := make([][2]int, len(prices))
	dp[0][0] = 0
	dp[0][1] = -prices[0]

	for i := 1; i < len(prices); i++ {
		dp[i][0] = Max(dp[i-1][0], dp[i-1][1]+prices[i])
		dp[i][1] = Max(dp[i-1][0]-prices[i], dp[i-1][1])
	}
	return dp[len(prices)-1][0]
}

func Max(x, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

// @lc code=end
